sources/packages/backend/apps/queue-consumers/src/services/cas-invoice-batch/cas-invoice-batch.service.ts

```diff
@@ -0,0 +1,254 @@
+      invoiceBatch.creator = this.systemUsersService.systemUser;
+      // Save the invoice batch to allow the ID to be known and associate with every new
+      // invoice created. It would allow the invoices to be saved in chunks.
+      await entityManager.getRepository(CASInvoiceBatch).save(invoiceBatch);
+      // Accumulates the invoices to be returned later.
+      const casInvoices: CASInvoice[] = [];
+      // Get active distribution accounts to be associated with each invoice detail.
+      const distributionAccounts = await this.getActiveDistributionAccounts(
+        entityManager,
+      );
+      await this.sequenceControlService.consumeNextSequenceWithExistingEntityManager(
+        "CAS_INVOICE",
+        entityManager,
+        async (nextSequenceNumber: number) => {
+          let newInvoiceSequence = nextSequenceNumber;
+          const invoicesSummary = new ProcessSummary();
+          processSummary.children(invoicesSummary);
+          // Start processing the invoices for each pending receipt.
+          for (const receipt of pendingReceipts) {
+            invoicesSummary.info(
+              `Creating invoice for receipt ID ${receipt.id}.`,
+            );
+            const student =
+              receipt.disbursementSchedule.studentAssessment.application
+                .student;
+            const offeringIntensity =
+              receipt.disbursementSchedule.studentAssessment.offering
+                .offeringIntensity;
+            // Create a new invoice for the receipt.
+            const newInvoice = new CASInvoice();
+            newInvoice.casInvoiceBatch = invoiceBatch;
+            newInvoice.disbursementReceipt = receipt;
+            newInvoice.casSupplier = student.casSupplier;
+            newInvoice.invoiceNumber = `SIMS-INVOICE-${newInvoiceSequence++}-${
+              student.casSupplier.supplierNumber
+            }`;
+            newInvoice.invoiceStatus = CASInvoiceStatus.Pending;
+            newInvoice.invoiceStatusUpdatedOn = now;
+            newInvoice.creator = this.systemUsersService.systemUser;
+            // Create invoice details.
+            newInvoice.casInvoiceDetails = this.createInvoiceDetails(
+              receipt.disbursementSchedule.disbursementValues,
+              offeringIntensity,
+              distributionAccounts,
+            );
+            casInvoices.push(newInvoice);
+            invoicesSummary.info(
+              `Invoice ${newInvoice.invoiceNumber} created for receipt ID ${receipt.id}.`,
+            );
+            newInvoice.casInvoiceDetails.forEach((detail) => {
+              invoicesSummary.info(
+                `Created invoice detail for award ${detail.casDistributionAccount.awardValueCode}(${detail.casDistributionAccount.operationCode}).`,
+              );
+            });
+          }
+          return newInvoiceSequence;
+        },
+      );
+      await entityManager.getRepository(CASInvoice).save(casInvoices, {
```

Comment:
Why we didn't prefer insert here? I could have improved the performance from what we have seen in load test work. We might have to manage the chunk manually in that case.

Reply:
Please take a closer look at the PR description. The save has performance issues when it executes a select to check if the record must be inserted or updated.
In the tests executed, no select was executed and the inserts were grouped, reducing drastically the "travels" to DB.

Reply:
I see the point now. when the primary key is not set, I believe then it ignores that overhead. Sounds good to me.